src/System.Private.CoreLib/src/System/Array.Portable.cs

```diff
@@ -0,0 +1,1659 @@
+                {
+                    Object obj = array.GetValue(i);
+                    if (obj == null)
+                    {
+                        if (value == null) return i;
+                    }
+                    else
+                    {
+                        if (obj.Equals(value)) return i;
+                    }
+                }
+            }
+            return lb - 1;
+        }
+
+        /// <summary>
+        /// This version is called from Array<T>.IndexOf and Contains<T>, so it's in every unique array instance due to array interface implementation.
+        /// Do not call into IndexOf<T>(Array array, Object value, int startIndex, int count) for size and space reasons.
+        /// Otherwise there will be two IndexOf methods for each unique array instance, and extra parameter checking which are not needed for the common case.
+        /// </summary>
+        public static int IndexOf<T>(T[] array, T value)
+        {
+            if (array == null)
+            {
+                throw new ArgumentNullException(nameof(array));
+            }
+
+            // See comment above Array.GetComparerForReferenceTypesOnly for details
+            EqualityComparer<T> comparer = GetComparerForReferenceTypesOnly<T>();
+
+            if (comparer != null)
+            {
+                for (int i = 0; i < array.Length; i++)
+                {
+                    if (comparer.Equals(array[i], value))
+                        return i;
+                }
+            }
+            else
+            {
+                for (int i = 0; i < array.Length; i++)
+                {
+                    if (StructOnlyEquals<T>(array[i], value))
+                        return i;
+                }
+            }
+
+            return -1;
+        }
+
+        public static int IndexOf<T>(T[] array, T value, int startIndex)
+        {
+            if (array == null)
+            {
+                throw new ArgumentNullException(nameof(array));
+            }
+
+            return IndexOf(array, value, startIndex, array.Length - startIndex);
+        }
+
+        public static int IndexOf<T>(T[] array, T value, int startIndex, int count)
+        {
+            if (array == null)
+            {
+                throw new ArgumentNullException(nameof(array));
+            }
+
+            if (startIndex < 0 || startIndex > array.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(startIndex), SR.ArgumentOutOfRange_Index);
+            }
+
+            if (count < 0 || count > array.Length - startIndex)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_Count);
+            }
+
+            int endIndex = startIndex + count;
+
+            // See comment above Array.GetComparerForReferenceTypesOnly for details
+            EqualityComparer<T> comparer = GetComparerForReferenceTypesOnly<T>();
```

Comment:
Places that call GetComparerForReferenceTypesOnly rely on  deep compiler magic that is not even portable between our two CoreRT compilers right now.